Source/WebCore/dom/InternalObserverForEach.cpp

```diff
@@ -0,0 +1,137 @@
+#include "SubscribeOptions.h"
+#include "Subscriber.h"
+#include "SubscriberCallback.h"
+#include "VisitorCallback.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+
+namespace WebCore {
+
+class InternalObserverForEach final : public InternalObserver {
+public:
+    static Ref<InternalObserverForEach> create(ScriptExecutionContext& context, Ref<VisitorCallback>&& callback, Ref<AbortSignal>&& signal, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverForEach(context, WTFMove(callback), WTFMove(signal), WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        auto* globalObject = protectedScriptExecutionContext()->globalObject();
+        ASSERT(globalObject);
+
+        Ref vm = globalObject->vm();
+
+        {
+            JSC::JSLockHolder lock(vm);
+
+            // The exception is not reported, instead it is forwarded to the
+            // abort signal and promise rejection.
+            // As such, VisitorCallback `[RethrowsException]` and here a
+            // catch scope is declared so the error can be passed to any promise
+            // rejection handlers and abort handlers.
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            Ref { m_callback }->handleEvent(value, m_idx++);
+
+            JSC::Exception* exception = scope.exception();
+            if (UNLIKELY(exception)) {
+                scope.clearException();
+                auto value = exception->value();
+                Ref { m_promise }->reject<IDLAny>(value);
+                Ref { m_signal }->signalAbort(value);
+            }
+        }
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        Ref { m_promise }->reject<IDLAny>(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+        Ref { m_promise }->resolve();
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor& visitor) const final
+    {
+        Ref { m_callback }->visitJSFunction(visitor);
+    }
+
+    void visitAdditionalChildren(JSC::SlotVisitor& visitor) const final
+    {
+        Ref { m_callback }->visitJSFunction(visitor);
+    }
+
+    InternalObserverForEach(ScriptExecutionContext& context, Ref<VisitorCallback>&& callback, Ref<AbortSignal>&& signal, Ref<DeferredPromise>&& promise)
+        : InternalObserver(context)
+        , m_callback(WTFMove(callback))
+        , m_signal(WTFMove(signal))
+        , m_promise(WTFMove(promise))
+    {
+    }
+
+    uint64_t m_idx { 0 };
+    Ref<VisitorCallback> m_callback;
+    Ref<AbortSignal> m_signal;
+    Ref<DeferredPromise> m_promise;
+};
+
+void createInternalObserverOperatorForEach(ScriptExecutionContext& context, Observable& observable, Ref<VisitorCallback>&& callback, const SubscribeOptions& options, Ref<DeferredPromise>&& promise)
+{
+    Ref signal = AbortSignal::create(&context);
```

Comment:
As per https://bugs.webkit.org/show_bug.cgi?id=282069, save's coming back to this later.